cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java

```diff
@@ -0,0 +1,191 @@
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Implements operations on {@link BusinessMetadataDataset} transactionally.
+ */
+public class BusinessMetadataStore {
+
+  private static final Id.DatasetInstance BUSINESS_METADATA_INSTANCE_ID =
+    Id.DatasetInstance.from(Id.Namespace.SYSTEM, "business.metadata");
+  private final Transactional<BusinessMdsIterable, BusinessMetadataDataset> txnl;
+
+  @Inject
+  public BusinessMetadataStore(TransactionExecutorFactory txExecutorFactory, final DatasetFramework dsFramework) {
+    this.txnl = Transactional.of(txExecutorFactory, new Supplier<BusinessMdsIterable>() {
+      @Override
+      public BusinessMdsIterable get() {
+        try {
+          Object dataset = DatasetsUtil.getOrCreateDataset(dsFramework, BUSINESS_METADATA_INSTANCE_ID,
+                                                            BusinessMetadataDataset.class.getSimpleName(),
+                                                            DatasetProperties.EMPTY, null, null);
+          if (dataset instanceof BusinessMetadataDataset) {
+            return new BusinessMdsIterable((BusinessMetadataDataset) dataset);
+          }
+          return new BusinessMdsIterable(new BusinessMetadataDataset((IndexedTable) dataset));
+        } catch (DatasetManagementException | IOException | ServiceUnavailableException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    });
+  }
+
+  /**
+   * Adds/updates metadata for the specified {@link Id.NamespacedId}.
+   *
+   * @param entityId the {@link Id.NamespacedId} for which metadata is to be updated
+   * @param metadata metadata (represented as a map) to update
+   */
+  void addMetadata(final Id.NamespacedId entityId, final Map<String, String> metadata) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        for (Map.Entry<String, String> entry : metadata.entrySet()) {
+          input.businessMds.createBusinessMetadata(entityId, entry.getKey(), entry.getValue());
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Adds tags for the specified {@link Id.NamespacedId}.
+   */
+  void addTags(final Id.NamespacedId entityId, final String ... tagsToAdd) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        Map<String, String> businessMetadata = input.businessMds.getBusinessMetadata(entityId);
+        Iterable<String> existingTags = getTags(businessMetadata);
+        Iterable<String> newTags = Iterables.concat(existingTags, Arrays.asList(tagsToAdd));
+        input.businessMds.createBusinessMetadata(entityId, "tags", Joiner.on(",").join(newTags));
```

Comment:
What happen if "tag" has comma? Do we check for it? Or would it be better using some safer encoding, e.g. Json?